lib/ttfunk/table/cff/charset.rb

```diff
@@ -0,0 +1,230 @@
+
+        def offset
+          # Numbers from 0..2 mean charset IDs instead of offsets. IDs are
+          # basically pre-defined sets of characters.
+          #
+          # In the case of an offset, add the CFF table's offset since the
+          # charset offset is relative to the start of the CFF table. Otherwise
+          # return nil (no offset).
+          if offset_or_id > 2
+            offset_or_id + top_dict.cff_offset
+          end
+        end
+
+        # mapping is new -> old glyph ids
+        def encode(mapping)
+          # no offset means no charset was specified (i.e. we're supposed to
+          # use a predefined charset) so there's nothing to encode
+          return '' unless offset
+
+          sids = mapping.keys.sort.map { |new_gid| sid_for(mapping[new_gid]) }
+          ranges = TTFunk::BinUtils.rangify(sids)
+          range_max = ranges.map(&:last).max
+
+          range_bytes = if range_max > 0
+                          (Math.log2(range_max) / 8).floor + 1
+                        else
+                          # for cases when there are no sequences at all
+                          Float::INFINITY
+                        end
+
+          # calculate whether storing the charset as a series of ranges is
+          # more efficient (i.e. takes up less space) vs storing it as an
+          # array of SID values
+          total_range_size = (2 * ranges.size) + (range_bytes * ranges.size)
+          total_array_size = sids.size * element_width(:array_format)
+
+          [].tap do |result|
+            if total_array_size <= total_range_size
+              result << [format_int(:array_format)].pack('C')
+              result << sids.pack('n*')
+            else
+              fmt = range_bytes == 1 ? :range_format_8 : :range_format_16
+              element_fmt = element_format(fmt)
+              result << [format_int(fmt)].pack('C')
+
+              ranges.each do |range|
+                sid, num_left = range
+                result << [sid, num_left].pack(element_fmt)
+              end
+            end
+          end.join
+        end
+
+        private
+
+        def sid_for(glyph_id)
+          return 0 if glyph_id == 0
+
+          # rather than validating the glyph as part of one of the predefined
+          # charsets, just pass it through
+          return glyph_id unless offset
+
+          case format_sym
+          when :array_format
+            # zero is always .notdef, so adjust with - 1
+            @entries[glyph_id - 1]
+
+          when :range_format_8, :range_format_16
+            remaining = glyph_id
+
+            @entries.each do |range|
+              if range.size >= remaining
+                return (range.first + remaining) - 1
+              end
+
+              remaining -= range.size
+            end
+
+            0
+          end
+        end
+
+        def find_string(sid)
+          if offset
+            return self.class.standard_strings[sid - 1] if sid <= 390
```

Comment:
Hmm that might be a good idea, hopefully I can avoid additional off-by-one issues when adding it to the various encodings 😅

Reply:
This one was pretty difficult to address because the fix isn't as simple as adding `.notdef` to each predefined charset. As it happens, the strings index also suffers from off by one issues because it too excludes GID 0. Since the strings index isn't an array though, prepending `.notdef` isn't easy to do like it is for the predefined charsets. The documentation assumes you understand the implicitness of GID 0 and would I think be more difficult to square up with what's in the code if we made allowances by inserting `.notdef` ourselves. That said, I do take your point that the many small adjustments are more prone to error, so I created a couple of classes to help abstract them away. I realized the majority of the discrepancies come from the fact that the predefined charsets and string index are really just 1-based arrays. The new class `OneBasedArray` and `OneBasedIndex` wrap `Array` and `CFF::Index` respectively and allow us to access elements directly by GID. Let me know what you think :)